src/v/io/io_queue.cc

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "io/io_queue.h"
+
+#include "io/logger.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sleep.hh>
+
+namespace experimental::io {
+
+io_queue::io_queue(
+  persistence* storage,
+  std::filesystem::path path,
+  completion_callback_type complete) noexcept
+  : storage_(storage)
+  , path_(std::move(path))
+  , complete_(std::move(complete)) {}
+
+void io_queue::start() noexcept {
+    log.debug("Starting io queue for {}", path_);
+    dispatcher_ = dispatch();
+}
+
+seastar::future<> io_queue::stop() noexcept {
+    log.debug("Stopping io queue for {}", path_);
+    auto dispatcher = std::exchange(
+      dispatcher_, seastar::make_ready_future<>());
+    stop_ = true;
+    cond_.signal();
+    co_await std::move(dispatcher);
+    co_await gate_.close();
+}
+
+const std::filesystem::path& io_queue::path() const noexcept { return path_; }
+
+seastar::future<> io_queue::open() noexcept {
```

Comment:
It's a bit strange that open/close methods are returning futures but submit_read/submit_write are not.

Reply:
Open and close perform blocking system calls so their return of future makes sense. However, submit_read/submit_write only queue operations. The intention here is that resource management such as applying back pressure is done at a higher level through mechanisms like memory reservations. It's entirely possible that we'll discover that we need to make submit_* return future at some point.